[SPARK-26859][SQL] Fix field writer index bug in non-vectorized ORC deserializer

IvanVergiliev opened the pull request:
## What changes were proposed in this pull request?

This happens in a schema evolution use case only when a user specifies the schema manually and use non-vectorized ORC deserializer code path.

There is a bug in `OrcDeserializer.scala` that results in `null`s being set at the wrong column position, and for state from previous records to remain uncleared in next records. There are more details for when exactly the bug gets triggered and what the outcome is in the [JIRA issue](https://jira.apache.org/jira/browse/SPARK-26859).

The high-level summary is that this bug results in severe data correctness issues, but fortunately the set of conditions to expose the bug are complicated and make the surface area somewhat small.

This change fixes the problem and adds a respective test.

## How was this patch tested?

Pass the Jenkins with the newly added test cases.

IvanVergiliev committed:
[sql] Fix data correctness bug in ORC deserializer

IvanVergiliev committed:
Simplify test

IvanVergiliev committed:
Move new test at the end of suite

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

dongjoon-hyun commented:
ok to test

SparkQA commented:
**[Test build #102263 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102263/testReport)** for PR 23766 at commit [`90bd80e`](https://github.com/apache/spark/commit/90bd80e5d95db2307c0027df6992ac2d64dc82e4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7868/
Test PASSed.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala:
```diff
@@ -47,18 +47,20 @@ class OrcDeserializer(
+      if (requestedColIds(targetColumnIndex) != -1) {
+        val value = orcStruct.getFieldValue(requestedColIds(targetColumnIndex))
+        if (value == null) {
+          resultRow.setNullAt(targetColumnIndex)
```
Good catch!

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala:
```diff
@@ -332,6 +334,24 @@ abstract class OrcSuite extends OrcTest with BeforeAndAfterAll {
   }
+
+  test("[SPARK-26859] Reading ORC files with explicit schema can result in wrong data") {
+    withSQLConf(SQLConf.ORC_VECTORIZED_READER_ENABLED.key -> "false") {
```
We need to disable vectorized orc to test this bug?

IvanVergiliev replied:
@viirya the `OrcDeserializer` class (which is the one that contains the bug) is only used for the non-vectorized reader AFAICT, so the vectorized doesn't go through the buggy codepath.

viirya replied:
I can't easily examine the codebase for now to check it, so thanks for explaining it.

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala:
```diff
@@ -332,6 +334,24 @@ abstract class OrcSuite extends OrcTest with BeforeAndAfterAll {
     }
   }
+
+  test("[SPARK-26859] Reading ORC files with explicit schema can result in wrong data") {
```
nit: we usually don't need bracket here. just `test("SPARK-26859: ...")`.

SparkQA commented:
**[Test build #102263 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102263/testReport)** for PR 23766 at commit [`90bd80e`](https://github.com/apache/spark/commit/90bd80e5d95db2307c0027df6992ac2d64dc82e4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102263/
Test PASSed.

dongjoon-hyun commented:
Hi, @IvanVergiliev .
Thank you for the fix. Apache Spark has a set of schema evolution test suite.
This should be added there, too. I made a [PR](https://github.com/IvanVergiliev/spark/pull/1) to you. Please review and merge that.

dongjoon-hyun reviewed (changes requested):
Hi, @IvanVergiliev .
Thank you for the fix. This is a schema evolution case and Apache Spark has a set of schema evolution test suite. We had better add into that, too. I made a [PR](https://github.com/IvanVergiliev/spark/pull/1) to you. Please review and merge that. Also, I applied @viirya 's review comments.

Also, I updated the PR title and `How was this patch tested?` section in the PR description.

dongjoon-hyun committed:
Add into schema evolution tests. (#1)

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7890/
Test PASSed.

IvanVergiliev commented:
@dongjoon-hyun @viirya thanks for the quick review!

> This is a schema evolution case and Apache Spark has a set of schema evolution test suite. We had better add into that, too. I made a [PR](https://github.com/IvanVergiliev/spark/pull/1) to you. Please review and merge that.

Oh cool, I wasn't aware of that. Thanks for incorporating the test there as well! I merged your changes into the branch.

SparkQA commented:
**[Test build #102287 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102287/testReport)** for PR 23766 at commit [`893fd7e`](https://github.com/apache/spark/commit/893fd7e15cd51d4bf3d0bef0b390d6250656eaa7).

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaTest.scala:
```diff
@@ -108,6 +108,40 @@ trait AddColumnTest extends ReadSchemaTest {
   }
 }
 
+trait AddColumnIntoTheMiddleTest extends ReadSchemaTest {
```
Can we also add few lines of comment for this trait as `AddColumnTest`?

dongjoon-hyun replied:
Oops. I missed that. Thanks.

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaSuite.scala:
```diff
@@ -163,6 +168,7 @@ class VectorizedParquetReadSchemaSuite
 
 class MergedParquetReadSchemaSuite
   extends ReadSchemaSuite
+  with AddColumnIntoTheMiddleTest
```
If we want to add `AddColumnIntoTheMiddleTest` to all of above, maybe we can just let `ReadSchemaSuite` extend `AddColumnIntoTheMiddleTest`.

dongjoon-hyun replied:
Hi, @viirya . `CSV` doesn't support this.

SparkQA commented:
**[Test build #102287 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102287/testReport)** for PR 23766 at commit [`893fd7e`](https://github.com/apache/spark/commit/893fd7e15cd51d4bf3d0bef0b390d6250656eaa7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait AddColumnIntoTheMiddleTest extends ReadSchemaTest `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102287/
Test PASSed.

dongjoon-hyun commented:
Hi, @IvanVergiliev . Could you update the followings?

* Update the existing one from `Case 1` to `Case 1-1`.
```
- * Add column (Case 1).
+ * Add column (Case 1-1).
```

* Add new comment as @viirya suggested.
```
+ /**
+  * Add column into the middle (Case 1-2).
+  */
```

dongjoon-hyun commented:
Also, cc @cloud-fan .

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/ReadSchemaTest.scala:
```diff
@@ -108,6 +108,40 @@ trait AddColumnTest extends ReadSchemaTest {
+      df2.write.format(format).options(options).save(dir2)
+
+      val df = spark.read
+        .schema(df2.schema)
```
just for curiosity, if we don't specify the schema, can spark infer the schema correctly?

IvanVergiliev replied:
If memory serves me well, schema merging is not supported for ORC. There’s an open issue for implementing it.

The current behavior is that the schema of the first file will be picked, and the rest will be ignored.

IvanVergiliev replied:
For future reference:

- [Support merge schema for ORC](https://issues.apache.org/jira/browse/SPARK-11412)
- [Feature parity for ORC with Parquet](https://issues.apache.org/jira/browse/SPARK-20901)

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala:
```diff
@@ -332,6 +334,23 @@ abstract class OrcSuite extends OrcTest with BeforeAndAfterAll {
     }
   }
+
+  test("SPARK-26859 Reading ORC files with explicit schema can result in wrong data") {
```
does this test help the test coverage? It seems is same as the `AddColumnIntoTheMiddleTest`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala:
```diff
@@ -47,18 +47,20 @@ class OrcDeserializer(
-        fieldWriters(i)(value)
+    var fieldWriterIndex = 0
+    var targetColumnIndex = 0
+    while (targetColumnIndex < requestedColIds.length) {
```
now this loop looks a little complicated, as there are 2 indexes.

How about we make the `fieldWriters` has the same size as `requiredSchema`? i.e.
```
private val fieldWriters: Array[WritableComparable[_] => Unit] = requiredSchema.zipWithIndex.map {
  case (f, index) =>
    if (requestedColIds(index) == -1) {
      null
    } else {
      // create writer
    }
}
```

Then this loop becomes
```
var i = 0
while (i < fieldWriters.length) {
  ...
}
```

IvanVergiliev replied:
I considered this as well. The currently implemented version allowed me to keep the change more isolated (a single method only), which seemed less risky and easier to review.

However, I agree that changing `fieldWriters` as well is probably better for readability in the long run. I’m definitely open to implementing it this way if there’s consensus that this is better.

(Minor note: if we do switch to that implementation, I would make the missing field writers None instead of null.)

cloud-fan replied:
In performance critical code path, we should prefer the java style, i.e. using null instead of `Option` to avoid overhead.

IvanVergiliev committed:
Rewrite OrcDeserializer.deserialize to use a single index only

IvanVergiliev committed:
Update Schema Evolution test comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7972/
Test PASSed.

IvanVergiliev commented:
- Updated the implementation to keep a `fieldWriter` or null for each target column so that the deserialization loop can use a single index;
- Updated the schema evolution comments.

Let me know if there's anything else that would be useful.

SparkQA commented:
**[Test build #102389 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102389/testReport)** for PR 23766 at commit [`5c59fff`](https://github.com/apache/spark/commit/5c59fff106914555ecdc6e719a474f6372e55f6e).

cloud-fan commented:
LGTM except one comment: https://github.com/apache/spark/pull/23766/files#r256712722

cloud-fan commented:
LGTM except one comment: https://github.com/apache/spark/pull/23766/files#r256712722

SparkQA commented:
**[Test build #102389 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102389/testReport)** for PR 23766 at commit [`5c59fff`](https://github.com/apache/spark/commit/5c59fff106914555ecdc6e719a474f6372e55f6e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102389/
Test FAILed.

IvanVergiliev commented:
The test failures appear relevant; I’ll take a look in a bit.

IvanVergiliev committed:
Fix length mismatch bug

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7978/
Test PASSed.

IvanVergiliev commented:
I figured it out. I was under the impression that `requiredSchema` and `requestedColIds` will have the same length. Apparently that's a wrong assumption - IIUC that's because the partition column is appended to the `requestedColIds` but not the schema.

I changed the code to only iterate until `fieldWriters.length` which should preserve existing behaviour since the columns after that, if any, wouldn't have field writers associated anyway.

SparkQA commented:
**[Test build #102398 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102398/testReport)** for PR 23766 at commit [`afe4b01`](https://github.com/apache/spark/commit/afe4b0186dd5601446f193c331dcda71c5f2821d).

SparkQA commented:
**[Test build #102398 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102398/testReport)** for PR 23766 at commit [`afe4b01`](https://github.com/apache/spark/commit/afe4b0186dd5601446f193c331dcda71c5f2821d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102398/
Test FAILed.

IvanVergiliev commented:
This one appears to be unrelated - I don't see `ORC` mentioned anywhere in the stack trace, and `OrcDeserializer` is the only non-test change in this PR.

IvanVergiliev committed:
Remove redundant test

IvanVergiliev commented:
@cloud-fan removed redundant test.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7980/
Test PASSed.

SparkQA commented:
**[Test build #102400 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102400/testReport)** for PR 23766 at commit [`2c7ce36`](https://github.com/apache/spark/commit/2c7ce3669371fd28f68927623c03fba72dd6d0d4).

SparkQA commented:
**[Test build #102400 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102400/testReport)** for PR 23766 at commit [`2c7ce36`](https://github.com/apache/spark/commit/2c7ce3669371fd28f68927623c03fba72dd6d0d4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102400/
Test PASSed.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/orc/OrcSourceSuite.scala:
```diff
@@ -35,6 +35,8 @@ import org.apache.spark.SPARK_VERSION_SHORT
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.test.SharedSQLContext
+import org.apache.spark.sql.types.{LongType, StringType, StructType}
+import org.apache.spark.sql.types.DataTypes._
```
Could you remove these changes because we removes the test cases?

IvanVergiliev replied:
Good point, removed. I'm surprised the linter/compiler didn't complain about these unused imports, it seems pretty strict otherwise.

IvanVergiliev committed:
Remove unnecessary imports

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8043/
Test PASSed.

SparkQA commented:
**[Test build #102467 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102467/testReport)** for PR 23766 at commit [`50de999`](https://github.com/apache/spark/commit/50de9995d2447da5242b1b26e2a70ab30b9a2439).

SparkQA commented:
**[Test build #102467 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102467/testReport)** for PR 23766 at commit [`50de999`](https://github.com/apache/spark/commit/50de9995d2447da5242b1b26e2a70ab30b9a2439).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102467/
Test FAILed.

cloud-fan commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8053/
Test PASSed.

SparkQA commented:
**[Test build #102480 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102480/testReport)** for PR 23766 at commit [`50de999`](https://github.com/apache/spark/commit/50de9995d2447da5242b1b26e2a70ab30b9a2439).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8054/
Test PASSed.

SparkQA commented:
**[Test build #102480 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102480/testReport)** for PR 23766 at commit [`50de999`](https://github.com/apache/spark/commit/50de9995d2447da5242b1b26e2a70ab30b9a2439).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102480/
Test FAILed.

cloud-fan commented:
retest this please

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcDeserializer.scala:
```diff
@@ -37,28 +37,34 @@ class OrcDeserializer(
+    var targetColumnIndex = 0
+    while (targetColumnIndex < fieldWriters.length) {
+      if (fieldWriters(targetColumnIndex) != null) {
+        var value = orcStruct.getFieldValue(requestedColIds(targetColumnIndex))
```
`value` can be `val` instead of `var`.

IvanVergiliev committed:
Make value a val instead of var

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8072/
Test PASSed.

SparkQA commented:
**[Test build #102501 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102501/testReport)** for PR 23766 at commit [`9059e02`](https://github.com/apache/spark/commit/9059e020af595f5f149c4f089cd215e0b670cb27).

SparkQA commented:
**[Test build #102501 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102501/testReport)** for PR 23766 at commit [`9059e02`](https://github.com/apache/spark/commit/9059e020af595f5f149c4f089cd215e0b670cb27).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102501/
Test FAILed.

cloud-fan commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8092/
Test PASSed.

SparkQA commented:
**[Test build #102528 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102528/testReport)** for PR 23766 at commit [`9059e02`](https://github.com/apache/spark/commit/9059e020af595f5f149c4f089cd215e0b670cb27).

SparkQA commented:
**[Test build #102528 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102528/testReport)** for PR 23766 at commit [`9059e02`](https://github.com/apache/spark/commit/9059e020af595f5f149c4f089cd215e0b670cb27).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102528/
Test FAILed.

dongjoon-hyun commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8094/
Test PASSed.

dongjoon-hyun reviewed (approved):
+1, LGTM.

SparkQA commented:
**[Test build #102532 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102532/testReport)** for PR 23766 at commit [`9059e02`](https://github.com/apache/spark/commit/9059e020af595f5f149c4f089cd215e0b670cb27).

SparkQA commented:
**[Test build #102532 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102532/testReport)** for PR 23766 at commit [`9059e02`](https://github.com/apache/spark/commit/9059e020af595f5f149c4f089cd215e0b670cb27).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102532/
Test FAILed.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #102541 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102541/testReport)** for PR 23766 at commit [`9059e02`](https://github.com/apache/spark/commit/9059e020af595f5f149c4f089cd215e0b670cb27).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/8105/
Test PASSed.

SparkQA commented:
**[Test build #102541 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102541/testReport)** for PR 23766 at commit [`9059e02`](https://github.com/apache/spark/commit/9059e020af595f5f149c4f089cd215e0b670cb27).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102541/
Test PASSed.

cloud-fan commented:
thanks, merging to master/2.4!

IvanVergiliev commented:
Awesome, thanks everyone for getting this reviewed!
